include/boost/math/special_functions/prime_functions.hpp

```diff
@@ -0,0 +1,106 @@
+// Copyright 2020 Matt Borland
+//
+// Use, modification and distribution are subject to the
+// Boost Software License, Version 1.0.
+// (See accompanying file LICENSE_1_0.txt
+// or copy at http://www.boost.org/LICENSE_1_0.txt)
+
+#ifndef BOOST_MATH_SPECIAL_FUNCTIONS_PRIME_FUNCTIONS_HPP
+#define BOOST_MATH_SPECIAL_FUNCTIONS_PRIME_FUNCTIONS_HPP
+
+#include <boost/math/special_functions/prime.hpp>
+#include <boost/assert.hpp>
+#include <deque>
+#include <vector>
+#include <iterator>
+
+namespace boost { namespace math
+{
+
+// https://mathworld.wolfram.com/SieveofEratosthenes.html
+// https://www.cs.utexas.edu/users/misra/scannedPdf.dir/linearSieve.pdf
+template<class Z, class OutputIterator>
+auto prime_sieve(Z lower_bound, Z upper_bound, OutputIterator output) -> decltype(output)
+{
+    static_assert(std::is_integral<Z>::value, "No primes for floating point types");
+    std::vector<Z> least_divisors(upper_bound + 1, 0);
```

Comment:
Is there a fast way to check if the sieve will overflows the type `Z`?

Reply:
@NAThompson I added an assertion for overflow to the new commit. It is not 0 cost, but it was the cheapest test I could think of.